hpctestlib/apps/amber/nve.py

```diff
@@ -0,0 +1,144 @@
+    is that Amber is already installed on the device under test.
+    '''
+
+    #: Amber output file.
+    #:
+    #: :default: : 'amber.out'
+    output_file = variable(str, value='amber.out')
+
+    #: Amber input file. This file is set by the post-init hook
+    #: :func:`unpack_platform_parameter`.
+    #:
+    #: :default: :class:`required`
+    input_file = variable(str)
+
+    #: Reference value of energy, that is used for the comparison
+    #: with the execution ouput on the sanity step. The absolute
+    #: difference between final energy value and reference value
+    #: should be smaller than energy_tolerance
+    #:
+    #: :type: float
+    #: :default: :class:`required`
+    energy_value = variable(float)
+
+    #: Maximum deviation from the reference value of energy,
+    #: that is acceptable.
+    #:
+    #: :type: float
+    #: :default: :class:`required`
+    energy_tolerance = variable(float)
+
+    #: Parameter pack containing the platform ID, input file and
+    #: executable.
+    platform_info = parameter([
+        ('cpu', 'mdin.CPU', 'pmemd.MPI'),
+        ('gpu', 'mdin.GPU', 'pmemd.cuda.MPI')
+    ])
+
+    #: NVE simulation parameter pack with the benchmark name,
+    #: energy reference and energy tolerance for each case.
+    variant = parameter([
+        ('Cellulose_production_NVE', -443246.0, 5.0E-05),
+        ('FactorIX_production_NVE', -234188.0, 1.0E-04),
+        ('JAC_production_NVE_4fs', -44810.0, 1.0E-03),
+        ('JAC_production_NVE', -58138.0, 5.0E-04)
+    ])
+
+    #: :default: :class:`required`
+    num_tasks_per_node = required
+
+    #: :default: :class:`required`
+    executable = required
+
+    tags = {'sciapp', 'chemistry'}
+
+    @run_after('init')
+    def unpack_platform_parameter(self):
+        '''Set the executable and input file.'''
+
+        self.platform, self.input_file, self.executable = self.platform_info
+
+    @run_after('init')
+    def unpack_variant_parameter(self):
+        '''Set the value of energy and energy tolerance for a
+        specific program.
+        '''
+
+        self.benchmark, self.energy_value, self.energy_tolerance = self.variant
+
+    @run_after('init')
+    def set_keep_files(self):
+        self.keep_files = [self.output_file]
+
+    @performance_function('ns/day', perf_key='nve')
+    def set_perf_patterns(self):
+        return sn.extractsingle(r'ns/day =\s+(?P<perf>\S+)',
+                                self.output_file, 'perf',
+                                float, item=1)
+
+    @run_before('performance')
+    def set_the_performance_dict(self):
+        self.perf_variables = {self.benchmark:
+                               sn.make_performance_function(
+                                   sn.extractsingle(
+                                       r'ns/day =\s+(?P<perf>\S+)',
+                                       self.output_file, 'perf',
+                                       float, item=1), 'ns/day')}
```

Comment:
Look. By default, `@performance_function` returns the dictionary as `perf_key: value`. In this case, we would need to define the reference value dictionary as `benchmark: nve: value`. This makes sense when `@performance_function`, but for this test it would result in an unnecessarily large and complex reference value structure. Therefore this function is added to change the final dictionary format.